src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -69,9 +71,21 @@ def resolve(self, root_reqs, check_supported_wheels):
         # The persistent state that we care about is `root_reqs`.
         assert len(self.factory.root_reqs) == 0, "Factory is being re-used"
 
-        # FIXME: Implement constraints.
-        if any(r.constraint for r in root_reqs):
-            raise InstallationError("Constraints are not yet supported.")
+        constraints = defaultdict(list)  # type: Dict[str,List[SpecifierSet]]
+        requirements = []
+        for req in root_reqs:
+            if req.constraint:
+                assert req.name
+                assert req.specifier
+                name = canonicalize_name(req.name)
+                constraints[name].append(req.specifier)
```

Comment:
I wonder if it is possible to merge the specifiers here. It would make the contraint checking code more readable since the loops can be eliminated.

Reply:
Hmm, I can look at that. Good suggestion.

Reply:
Yes we can. `SpecifierSet & SpecifierSet` does what we want. I'll fix this tomorrow. It's getting late and I don't want to mess things up, with the refactor PR being based on this one.